File: java/src/main/java/defpackage/PopupMenu.java
package defpackage;

import javax.microedition.lcdui.Graphics;

/* renamed from: af */
/* JADX INFO: loaded from: Heroes-Lore-Wind-of-Soltia_J2ME_EN_v207.jar:af.class */
/**
 * General-purpose popup / dialog spawned by {@link Menu#showPopup} and
 * {@link Menu#showMessage}. Its {@link #type} code selects the layout:
 * <ul>
 *   <li>1 / 11 — OK message (single softkey);</li>
 *   <li>2 / 6 / 12 — yes-no confirm;</li>
 *   <li>3 / 4 / 5 / 8 — a selectable option list (8 keeps row 0 as a header);</li>
 *   <li>9 — a bare message with no softkeys.</li>
 * </ul>
 * Message layouts draw {@link #joinedText} (all options joined by newlines);
 * list layouts draw {@link #options} row by row with the cursor arrow. The two
 * softkey labels {@link #positiveLabel} / {@link #negativeLabel} default per
 * type. Answers are reported back to the parent through
 * {@code onPopupResult(type, result)} where {@code result} is the chosen row,
 * {@code 0} for OK/Yes, or {@code 99} for cancel/Back. {@link #boxHeight} is the
 * panel height pre-measured in the constructor.
 */
public final class PopupMenu extends Menu {
    /* renamed from: c */
    /** Layout/behavior selector — see the class doc for the type codes. */
    private byte type;

    /* renamed from: a */
    /** The option rows as passed in (each element is a {@code char[]}). */
    private Object[] options;

    /* JADX INFO: renamed from: a, reason: collision with other field name */
    /** {@link #options} joined by newlines into one block, for message layouts. */
    private char[] joinedText;

    /* JADX INFO: renamed from: a, reason: collision with other field name */
    /** Pre-measured panel height for the current type and content. */
    private int boxHeight;

    /* renamed from: b */
    /** Left/confirm softkey label (OK/Yes/…), defaulted from {@link #type}. */
    private char[] positiveLabel;

    /* JADX INFO: renamed from: c, reason: collision with other field name */
    /** Right/cancel softkey label (Back/No/…), defaulted from {@link #type}. */
    private char[] negativeLabel;

    public PopupMenu(Menu parent, byte type, byte itemCount, Object[] options, char[] positiveLabel, char[] negativeLabel) {
        super(parent, itemCount);
        this.type = type;
        StringBuffer joined = new StringBuffer();
        for (int i = 0; i < options.length; i++) {
            char[] option = (char[]) options[i];
            if (option.length > 0) {
                joined.append(option);
                if (i != options.length - 1) {
                    joined.append('\n');
                }
            }
        }
        this.joinedText = joined.toString().toCharArray();
        this.options = options;
        if (type == 2 || type == 12) {
            positiveLabel = positiveLabel == null ? FontManager.labelYes : positiveLabel;
            if (negativeLabel == null) {
                negativeLabel = FontManager.labelNo;
            }
        } else if (type == 1 || type == 11) {
            if (positiveLabel == null) {
                positiveLabel = FontManager.labelOk;
            }
        } else if (type != 9) {
            positiveLabel = positiveLabel == null ? FontManager.labelOk : positiveLabel;
            if (negativeLabel == null) {
                negativeLabel = FontManager.labelBack;
            }
        }
        this.positiveLabel = positiveLabel;
        this.negativeLabel = negativeLabel;
        switch (type) {
            case 2:
            case 6:
                this.boxHeight = 8 + FontManager.measureBlockHeight(FontManager.percentOf(BaseCanvas.width, 80) - 10, 1, this.joinedText, 0, 0, this.joinedText.length);
                break;
            case 3:
            case 4:
            case 5:
            case 8:
                this.boxHeight = 12;
                for (int optionIndex = 0; optionIndex < this.options.length; optionIndex++) {
                    char[] option = (char[]) this.options[optionIndex];
                    this.boxHeight += 3 + FontManager.measureBlockHeight(FontManager.percentOf(BaseCanvas.width, 80) - 10, 1, option, 0, 0, option.length);
                }
                break;
            case 7:
            case 9:
            case 10:
            default:
                this.boxHeight = 22 + FontManager.measureBlockHeight(FontManager.percentOf(BaseCanvas.width, 80) - 10, 1, this.joinedText, 0, 0, this.joinedText.length);
                break;
            case 11:
            case 12:
                this.boxHeight = 8 + FontManager.measureBlockHeight(FontManager.percentOf(BaseCanvas.width, 80) - 10, 1, this.joinedText, 0, 0, this.joinedText.length);
                break;
        }
        if (type == 6) {
            ((Menu) this).cursorIndex = (byte) 1;
        }
    }

    @Override // defpackage.cb
    public final boolean handleKey(int action, int keyCode) {
        if (passKeyToChild(action, keyCode)) {
            return true;
        }
        switch (this.type) {
            case 1:
            case 11:
                if (keyCode == 53 || action == 8) {
                    ((Menu) this).parent.onPopupResult(this.type, (byte) 0);
                }
                break;
            case 2:
            case 6:
            case 12:
                if (keyCode == 53) {
                    ((Menu) this).parent.onPopupResult(this.type, (byte) 0);
                } else if (keyCode == -8) {
                    ((Menu) this).parent.onPopupResult(this.type, (byte) 99);
                }
                break;
            case 3:
            case 4:
            case 5:
            case 8:
                if (!moveCursorVerticalNoWrap(action, keyCode)) {
                    if (keyCode == 53 || action == 8) {
                        ((Menu) this).parent.onPopupResult(this.type, ((Menu) this).cursorIndex);
                    } else if (keyCode == -8) {
                        ((Menu) this).parent.onPopupResult(this.type, (byte) 99);
                    }
                    break;
                }
                break;
        }
        return true;
    }

    @Override // defpackage.cb
    public final void paint(Graphics graphics, int x, int y) {
        FontManager.clearScreen(graphics);
        int boxWidth = FontManager.percentOf(BaseCanvas.width, 80);
        int boxX = BaseCanvas.halfW - (boxWidth >> 1);
        int boxY = BaseCanvas.halfH - (this.boxHeight >> 1);
        Menu.drawPanelFrame(graphics, boxX, boxY, boxWidth, this.boxHeight);
        Menu.fillPanelInterior(graphics, boxX, boxY, boxWidth, this.boxHeight);
        switch (this.type) {
            case 1:
            case 9:
                graphics.setColor(16777215);
                FontManager.drawWrappedText(graphics, boxX + 5, boxY + 5, boxWidth - 10, 1, this.joinedText);
                if (this.type != 9) {
                    FontManager.drawSoftKeys(graphics, this.positiveLabel, (char[]) null);
                }
                break;
            case 2:
            case 6:
                graphics.setColor(16777215);
                FontManager.drawWrappedText(graphics, boxX + 5, boxY + 5, boxWidth - 10, 1, this.joinedText);
                break;
            case 3:
            case 4:
            case 5:
                graphics.setColor(16777215);
                int rowY = boxY + 7;
                for (byte optionIndex = 0; optionIndex < this.options.length; optionIndex = (byte) (optionIndex + 1)) {
                    if (optionIndex == ((Menu) this).cursorIndex) {
                        graphics.drawImage(AssetCache.cursorArrow, boxX + 5, rowY, 20);
                    }
                    rowY += 3 + FontManager.drawWrappedText(graphics, boxX + 12, rowY, boxWidth - 10, 1, (char[]) this.options[optionIndex]);
                }
                break;
            case 8:
                graphics.setColor(16777215);
                int headerY = boxY + 5;
                int listY = headerY + 3 + FontManager.drawWrappedText(graphics, boxX + 5, headerY, boxWidth - 10, 1, (char[]) this.options[0]);
                for (byte optionIndex2 = 1; optionIndex2 < this.options.length; optionIndex2 = (byte) (optionIndex2 + 1)) {
                    if (optionIndex2 == ((Menu) this).cursorIndex + 1) {
                        graphics.drawImage(AssetCache.cursorArrow, boxX + 5, listY, 20);
                    }
                    listY += 3 + FontManager.drawWrappedText(graphics, boxX + 12, listY, boxWidth - 10, 1, (char[]) this.options[optionIndex2]);
                }
                break;
            case 11:
                FontManager.drawWrappedText(graphics, boxX + 5, boxY + 5, boxWidth - 10, 1, this.joinedText);
                break;
            case 12:
                FontManager.drawWrappedText(graphics, boxX + 5, boxY + 5, boxWidth - 10, 1, this.joinedText);
                break;
        }
        FontManager.drawSoftKeys(graphics, this.positiveLabel, this.negativeLabel);
    }
}

File: transliteration/game-xlat/src/popup_menu.rs
//! Transliterated from `java/src/main/java/defpackage/PopupMenu.java`
//! (original `af.class` in `Heroes-Lore-Wind-of-Soltia_J2ME_EN_v207.jar`).
//!
//! Implementation #1: strict transliteration. See `docs/TRANSLITERATION.md`.
//!
//! General-purpose popup / dialog (`PopupMenu extends Menu`) spawned by
//! [`Menu::show_popup`](crate::menu::show_popup) / [`show_message`](crate::menu::show_message).
//! Its [`popup_type`](PopupMenuState::popup_type) selects the layout: `1`/`11` OK
//! message, `2`/`6`/`12` yes-no confirm, `3`/`4`/`5`/`8` selectable list (8 keeps
//! row 0 as a header), `9` a bare message. Answers are reported back to the parent
//! via `parent.onPopupResult(type, result)` — `result` is the chosen row, `0` for
//! OK/Yes, `99` for cancel/Back (see [`crate::menu::on_popup_result`]).
//!
//! ## ANTI-BOG boundary
//!
//! This increment ports the constructor (field stores + the `joinedText` build +
//! the label defaulting), `handleKey` (fully — the popup->parent
//! `onPopupResult` callback is made **real** via the flat model's
//! [`parent_of`](crate::menu::parent_of) scan), and a **DEFERRED** `paint`. The
//! `paint` is fully deferred: its very first op (`FontManager.clearScreen`) and the
//! box geometry (`FontManager.percentOf(BaseCanvas.width, 80)`) cross into unported
//! `FontManager` statics, and the body draws `FontManager.drawWrappedText` /
//! `drawSoftKeys` / `AssetCache.cursorArrow` — all unported. The `boxHeight`
//! pre-measurement (`FontManager.measureBlockHeight` / `percentOf`) is likewise
//! DEFERRED, and the `positiveLabel`/`negativeLabel` defaults that resolve to
//! `FontManager.labelYes`/`labelNo`/`labelBack` (unported) are DEFERRED — only the
//! `FontManager.labelOk` default (ported) is filled.
//!
//! `PopupMenu`'s fields are all per-INSTANCE (the `Menu` base fields likewise), so
//! it contributes no `java/reconstruction/ownership.tsv` static rows.
//!
//! Opcode shapes (R8, `_reference/numeric_shapes.json`):
//! `af.<init>:(Lcb;BB[Ljava/lang/Object;[C[C)V =>
//! [isub,iinc,isub,iadd,isub,iadd,isub,iadd,iadd,iinc,isub,iadd]` (constructor —
//! the ported subset builds `joinedText` (`i != options.length - 1` → the `isub`)
//! and stores; the remaining `iadd/isub`s live in the DEFERRED `boxHeight`
//! measurement), `af.a:(II)Z => []` (handleKey — pure branches), and
//! `af.a:(…Graphics;II)V => [ishr,isub,ishr,…]` (paint — fully DEFERRED).

use crate::game::Game;
use crate::menu::{self, MenuNode};

/// Java `af` / `PopupMenu` instance state — the `Menu` (`cb`) base fields plus the
/// popup's own instance fields.
#[derive(Debug, Default, Clone)]
pub struct PopupMenuState {
    /// The `Menu` (`cb`) base instance fields.
    pub base: menu::MenuBase,
    /// `private byte type;` (obf `c`) — layout/behaviour selector (`type` is a Rust
    /// keyword, so it is `popup_type` here).
    pub popup_type: i8,
    /// `private Object[] options;` (obf `a`) — the option rows (each a `char[]`).
    pub options: Vec<Vec<u16>>,
    /// `private char[] joinedText;` (obf `a`) — `options` joined by newlines.
    pub joined_text: Vec<u16>,
    /// `private int boxHeight;` (obf `a`) — pre-measured panel height (DEFERRED → 0).
    pub box_height: i32,
    /// `private char[] positiveLabel;` (obf `b`) — left/confirm softkey label.
    pub positive_label: Option<Vec<u16>>,
    /// `private char[] negativeLabel;` (obf `c`) — right/cancel softkey label.
    pub negative_label: Option<Vec<u16>>,
}

/// `public PopupMenu(Menu parent, byte type, byte itemCount, Object[] options, char[] positiveLabel, char[] negativeLabel)`
/// (`af.<init>:(Lcb;BB[Ljava/lang/Object;[C[C)V`). Builds `joinedText`, stores the
/// options, defaults the softkey labels per `type`, pre-measures `boxHeight`
/// (DEFERRED), and — for `type == 6` — starts the cursor on row 1.
pub fn construct(
    g: &mut Game,
    popup_type: i8,
    item_count: i8,
    options: Vec<Vec<u16>>,
    mut positive_label: Option<Vec<u16>>,
    negative_label: Option<Vec<u16>>,
) {
    // super(parent, itemCount);   (parent is the calling menu → non-null → present)
    g.popup_menu.base = menu::construct(true, item_count);
    // this.type = type;
    g.popup_menu.popup_type = popup_type;
    // StringBuffer joined = new StringBuffer();
    // for (int i = 0; i < options.length; i++) { char[] option = options[i];
    //   if (option.length > 0) { joined.append(option); if (i != options.length - 1) joined.append('\n'); } }
    let mut joined: Vec<u16> = Vec::new();
    let n = options.len();
    for (i, option) in options.iter().enumerate() {
        if !option.is_empty() {
            joined.extend_from_slice(option);
            if i != n - 1 {
                joined.push(b'\n' as u16);
            }
        }
    }
    // this.joinedText = joined.toString().toCharArray();
    g.popup_menu.joined_text = joined;
    // this.options = options;
    g.popup_menu.options = options;
    // --- label defaulting per type ---
    // (Only `positive_label` is reassigned here — the `negative_label` defaults resolve
    //  to `FontManager.labelNo`/`labelBack`, both unported → DEFERRED, so it stays as passed.)
    if popup_type == 2 || popup_type == 12 {
        // positiveLabel = positiveLabel == null ? FontManager.labelYes : positiveLabel;
        if positive_label.is_none() {
            // (DEFERRED: FontManager.labelYes unported — the null default is left None.)
        }
        // if (negativeLabel == null) negativeLabel = FontManager.labelNo;
        if negative_label.is_none() {
            // (DEFERRED: FontManager.labelNo unported.)
        }
    } else if popup_type == 1 || popup_type == 11 {
        // if (positiveLabel == null) positiveLabel = FontManager.labelOk;
        if positive_label.is_none() {
            positive_label = g.font_manager.label_ok.clone();
        }
    } else if popup_type != 9 {
        // positiveLabel = positiveLabel == null ? FontManager.labelOk : positiveLabel;
        if positive_label.is_none() {
            positive_label = g.font_manager.label_ok.clone();
        }
        // if (negativeLabel == null) negativeLabel = FontManager.labelBack;
        if negative_label.is_none() {
            // (DEFERRED: FontManager.labelBack unported.)
        }
    }
    // this.positiveLabel = positiveLabel; this.negativeLabel = negativeLabel;
    g.popup_menu.positive_label = positive_label;
    g.popup_menu.negative_label = negative_label;
    // switch (type) { ... this.boxHeight = ... measureBlockHeight(percentOf(width,80)-10, ...) ... }
    // (DEFERRED: boxHeight measurement — FontManager.measureBlockHeight / percentOf /
    //  BaseCanvas.width percent are unported; boxHeight stays 0.)
    // if (type == 6) ((Menu) this).cursorIndex = (byte) 1;
    if popup_type == 6 {
        g.popup_menu.base.cursor_index = 1;
    }
}

/// `public final boolean handleKey(int action, int keyCode)` (`af.a:(II)Z => []`):
/// forwards to any child, then dispatches by `type`. OK/Yes/list-select report
/// `parent.onPopupResult(type, result)` (`result` = `0` for OK/Yes, the row for a
/// list, `99` for Back). Always returns true.
// The `case 3/4/5/8` arm keeps the Java `if (!moveCursorVerticalNoWrap(...)) { … }`
// structure verbatim (the guard has the cursor-move side effect), so the
// `collapsible_match` suggestion to fold it into a match guard is declined.
#[allow(clippy::collapsible_match)]
pub fn handle_key(g: &mut Game, action: i32, key_code: i32) -> bool {
    // if (passKeyToChild(action, keyCode)) return true;
    if menu::pass_key_to_child(g, MenuNode::Popup, action, key_code) {
        return true;
    }
    // switch (this.type) {
    match g.popup_menu.popup_type as i32 {
        // case 1: case 11:
        1 | 11 => {
            // if (keyCode == 53 || action == 8) parent.onPopupResult(type, (byte) 0);
            if key_code == 53 || action == 8 {
                popup_result(g, 0);
            }
        }
        // case 2: case 6: case 12:
        2 | 6 | 12 => {
            // if (keyCode == 53) parent.onPopupResult(type, 0); else if (keyCode == -8) parent.onPopupResult(type, 99);
            if key_code == 53 {
                popup_result(g, 0);
            } else if key_code == -8 {
                popup_result(g, 99);
            }
        }
        // case 3: case 4: case 5: case 8:
        3 | 4 | 5 | 8 => {
            // if (!moveCursorVerticalNoWrap(action, keyCode)) {
            if !menu::move_cursor_vertical(&mut g.popup_menu.base, action, key_code, false) {
                // if (keyCode == 53 || action == 8) parent.onPopupResult(type, cursorIndex);
                if key_code == 53 || action == 8 {
                    let cursor = g.popup_menu.base.cursor_index;
                    popup_result(g, cursor);
                // else if (keyCode == -8) parent.onPopupResult(type, 99);
                } else if key_code == -8 {
                    popup_result(g, 99);
                }
            }
        }
        _ => {}
    }
    // return true;
    true
}

/// `((Menu) this).parent.onPopupResult(this.type, result)` — the flat model
/// resolves the parent via [`menu::parent_of`] (the popup is the current child of
/// exactly one menu) and dispatches the virtual `onPopupResult`.
fn popup_result(g: &mut Game, result: i8) {
    let tag = g.popup_menu.popup_type;
    // ((Menu) this).parent  — the menu whose child is this popup.
    let parent =
        menu::parent_of(g, MenuNode::Popup).expect("NullPointerException: PopupMenu.parent");
    menu::on_popup_result(g, parent, tag, result);
}

/// `public final void paint(Graphics graphics, int x, int y)`
/// (`af.a:(…Graphics;II)V`): **DEFERRED**. The full body crosses into unported
/// `FontManager` statics from its first statement onward (`clearScreen`,
/// `percentOf`, `drawWrappedText`, `drawSoftKeys`, `AssetCache.cursorArrow`), and
/// the panel geometry depends on the DEFERRED `percentOf`, so nothing portable
/// remains to draw. Faithful full form (per `type`):
///   FontManager.clearScreen(graphics);
///   int boxWidth = FontManager.percentOf(BaseCanvas.width, 80);
///   int boxX = BaseCanvas.halfW - (boxWidth >> 1);
///   int boxY = BaseCanvas.halfH - (this.boxHeight >> 1);
///   Menu.drawPanelFrame(graphics, boxX, boxY, boxWidth, this.boxHeight);
///   Menu.fillPanelInterior(graphics, boxX, boxY, boxWidth, this.boxHeight);
///   switch (type) { … drawWrappedText(joinedText) / list rows with AssetCache.cursorArrow … }
///   FontManager.drawSoftKeys(graphics, this.positiveLabel, this.negativeLabel);
pub fn paint(_g: &mut Game, _x: i32, _y: i32) {
    // (DEFERRED — see the doc comment. FontManager.clearScreen / percentOf /
    //  drawWrappedText / drawSoftKeys and AssetCache.cursorArrow are unported.)
}
